src/service/metrics/prom.rs

```diff
@@ -502,124 +502,165 @@ fn get_metadata_object(schema: &Schema) -> Option<prom::MetadataObject> {
     start: i64,
     end: i64,
 ) -> Result<Vec<String>> {
-    let stream_name = match selector {
-        Some(selector) => match selector.name {
-            Some(v) => v,
-            None => {
-                let labels = selector.matchers.find_matchers(NAME_LABEL);
-                if !labels.is_empty() {
-                    labels.first().unwrap().to_string()
-                } else {
-                    "".to_string()
-                }
-            }
-        },
-        None => "".to_string(),
+    let opt_metric_name = selector.and_then(try_into_metric_name);
+    let stream_schemas = match db::schema::list(org_id, Some(StreamType::Metrics), true).await {
+        Err(_) => return Ok(vec![]),
+        Ok(schemas) => schemas,
     };
-    let streams = db::schema::list(org_id, Some(StreamType::Metrics), true)
-        .await
-        .unwrap_or_default();
-    let mut labels = FxHashSet::default();
-    streams.iter().for_each(|v| {
-        let stats = stats::get_stream_stats(org_id, &v.stream_name, StreamType::Metrics);
-        if !stream_name.is_empty() && v.stream_name != stream_name {
-            return;
+    let mut label_names = FxHashSet::default();
+    for schema in stream_schemas {
+        if let Some(ref metric_name) = opt_metric_name {
+            if *metric_name != schema.stream_name {
+                // Client has requested a particular metric name, but this stream is
+                // not it.
+                continue;
+            }
         }
-        // check the stream is active, last updated >= start - file_push_interval && first updated >= end
-        if stats.doc_time_max
-            >= start
-                - Duration::seconds(CONFIG.limit.file_push_interval as i64)
-                    .num_microseconds()
-                    .unwrap()
-            && stats.doc_time_min < end
-        {
-            v.schema.fields().iter().for_each(|v| {
-                labels.insert(v.name().to_string());
-            });
+        let stats = stats::get_stream_stats(org_id, &schema.stream_name, StreamType::Metrics);
+        if stats.time_range_intersects(start, end) {
+            let field_names = schema
+                .schema
+                .fields()
+                .iter()
+                .map(|f| f.name())
+                .filter(|&s| {
+                    s != &CONFIG.common.column_timestamp && s != VALUE_LABEL && s != HASH_LABEL
+                })
+                .cloned();
+            label_names.extend(field_names);
         }
-    });
-    let mut labels = labels
-        .into_iter()
-        .filter(|v| v != &CONFIG.common.column_timestamp && v != HASH_LABEL && v != VALUE_LABEL)
-        .collect::<Vec<String>>();
-    labels.sort();
-    Ok(labels)
+    }
+    let mut label_names = label_names.into_iter().collect::<Vec<_>>();
+    label_names.sort();
+    Ok(label_names)
 }
 
+// XXX-TODO: filter the results in accordance with `selector.matchers`
```

Comment:
You are right, i forgot handle it

Reply:
this section i checked, it it no problem, because I already checked `metric name`, we don't need other matches, other labels just used for filter series, but labels are same for different series in one stream.